osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -0,0 +1,178 @@
+                                 .Where(kvp => kvp.Key.IsBonus())
+                                 .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value));
+
+            foreach (var result in bonusHits)
+                processor.ApplyResult(new JudgementResult(null!, new FakeJudgement(result)) { Type = result });
+
+            // Not true for all scores for whatever reason. Oh well.
+            // Debug.Assert(processor.HighestCombo.Value == score.MaxCombo);
+
+            return processor.TotalScore.Value;
+        }
+
+        private static HitResult getMaxJudgementFor(HitResult hitResult, HitResult max)
+        {
+            switch (hitResult)
+            {
+                case HitResult.Miss:
+                case HitResult.Meh:
+                case HitResult.Ok:
+                case HitResult.Good:
+                case HitResult.Great:
+                case HitResult.Perfect:
+                    return max;
+
+                case HitResult.SmallTickMiss:
+                case HitResult.SmallTickHit:
+                    return HitResult.SmallTickHit;
+
+                case HitResult.LargeTickMiss:
+                case HitResult.LargeTickHit:
+                    return HitResult.LargeTickHit;
+            }
+
+            return HitResult.IgnoreHit;
+        }
+
+        public static long GetOldStandardised(ScoreInfo score)
+        {
+            double accuracyScore =
+                (double)score.Statistics.Where(kvp => kvp.Key.AffectsAccuracy()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value)
+                / score.MaximumStatistics.Where(kvp => kvp.Key.AffectsAccuracy()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value);
+            double comboScore = (double)score.MaxCombo / score.MaximumStatistics.Where(kvp => kvp.Key.AffectsCombo()).Sum(kvp => kvp.Value);
+            double bonusScore = score.Statistics.Where(kvp => kvp.Key.IsBonus()).Sum(kvp => Judgement.ToNumericResult(kvp.Key) * kvp.Value);
+
+            double accuracyPortion = 0.3;
+
+            switch (score.RulesetID)
+            {
+                case 1:
+                    accuracyPortion = 0.75;
+                    break;
+
+                case 3:
+                    accuracyPortion = 0.99;
+                    break;
+            }
+
+            return (long)(1000000 * (accuracyPortion * accuracyScore + (1 - accuracyPortion) * comboScore) + bonusScore);
+        }
+
+        private class FakeHit : HitObject
+        {
+            private readonly Judgement judgement;
+
+            public override Judgement CreateJudgement() => judgement;
+
+            public FakeHit(Judgement judgement)
+            {
+                this.judgement = judgement;
+            }
+        }
+
+        private class FakeJudgement : Judgement
+        {
+            public override HitResult MaxResult { get; }
+
+            public FakeJudgement(HitResult result)
+            {
+                MaxResult = result;
+            }
```

Comment:
Maybe this is going to be considered overly pedantic in what is ostensibly temporary code but I do find myself quite baffled by this. Half the time `FakeJudgement` is constructed with a `MaxResult` being the actual max result of the hit result given (the times where `getMaxJudgementFor()` is called), and half the time it is just given the _actual_ result hit by the player and which becomes the max result, even if it is a miss? Quite confusing this, especially so that the ctor param is called `result` but is assigned to `MaxResult`, but there is also `Type`, and so I didn't understand why "worst case scenarios" with misses were doing anything with `getMaxJudgementFor()` given they're... misses.

If possible I'd like to inline the `getMaxJudgementFor()` machinery firmly inside this class and just have one ctor which takes in the actual result hit by the player, and the max judgement for the ruleset. `MaxResult` can be derived safely from both.

Reply:
It can be moved inside, as long as we want to completely ignore `MaximumStatistics` in the database. That's the weird case you're probably getting confused by right now.

There's argument for ignoring it *if* `getMaxJudgementFor` is correct in all cases. I only added this towards the end, when dealing with very old scores that could not use the `MaximumStatistics`.

Reply:
> as long as we want to completely ignore `MaximumStatistics` in the database. That's the weird case you're probably getting confused by right now.

Okay, that explains that. Probably fine to ignore this then...

Reply:
I've commented this better, hopefully.